File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/Location/MSLocationController.m
//
//  MSLocationController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/11/29.
//

#import "MSLocationController.h"
#import "MSLocationListCell.h"
#import "MSIMSDK-UIKit.h"
#import <MJRefresh.h>
#import <AMapFoundationKit/AMapFoundationKit.h>
#import <MAMapKit/MAMapKit.h>
#import <AMapSearchKit/AMapSearchKit.h>
#import <MSIMSDK/MSIMSDK.h>


@interface MSLocationController ()<UITableViewDelegate,UITableViewDataSource,AMapSearchDelegate,MAMapViewDelegate>

@property(nonatomic,strong) UIView *navView;

@property(nonatomic,strong) MAMapView *mapView;

@property(nonatomic,strong) AMapSearchAPI *searchAPI;

@property(nonatomic,strong) AMapPOIAroundSearchRequest *poiRequest;

@property(nonatomic,strong) CLLocation *userLocation;

@property(nonatomic,assign) NSInteger page;

@property(nonatomic,strong) UITableView *myTableView;

@property(nonatomic,strong) UIButton *myLocationBtn;

@property(nonatomic,strong) NSMutableArray<MSLocationInfo *> *listArr;

@property(nonatomic,strong) MSLocationInfo *selectInfo;

@property(nonatomic,strong) MAPointAnnotation *annotation;

@end

@implementation MSLocationController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor d_colorWithColorLight:TController_Background_Color dark:TController_Background_Color_Dark];

    [self setupNavView];
    [AMapServices sharedServices].apiKey = GaodeAPIKey;
    [AMapSearchAPI updatePrivacyShow:AMapPrivacyShowStatusDidShow privacyInfo:AMapPrivacyInfoStatusDidContain];
    [AMapSearchAPI updatePrivacyAgree:AMapPrivacyAgreeStatusDidAgree];
    [AMapServices sharedServices].enableHTTPS = YES;

    [self.view addSubview:self.mapView];
    [self.view addSubview:self.myTableView];
    [self.view addSubview:self.myLocationBtn];
}

- (void)dealloc
{
    MSLog(@"%@ dealloc",self.class);
}

static inline CGFloat kTableViewMaxY() {
    return Screen_Height*0.6;
}

static inline CGFloat kTableViewMinY() {
    return Screen_Height*0.4;
}

- (UITableView *)myTableView
{
    if (!_myTableView) {
        _myTableView = [[UITableView alloc]initWithFrame:CGRectMake(0, kTableViewMaxY(), Screen_Width, Screen_Height - kTableViewMaxY()) style:UITableViewStylePlain];
        _myTableView.dataSource = self;
        _myTableView.delegate = self;
        _myTableView.tableFooterView = [UIView new];
        _myTableView.backgroundColor = self.view.backgroundColor;
        [_myTableView registerClass:[MSLocationListCell class] forCellReuseIdentifier:@"locationCell"];
        _myTableView.rowHeight = 70;
        _myTableView.separatorColor = [UIColor d_colorWithColorLight:TCell_separatorColor dark:TCell_separatorColor_Dark];
        _myTableView.layer.cornerRadius = 16;
        _myTableView.clipsToBounds = YES;
        WS(weakSelf)
        _myTableView.mj_footer = [MJRefreshAutoNormalFooter footerWithRefreshingBlock:^{
            [weakSelf loadMorePOI];
        }];
    }
    return _myTableView;
}

- (MAMapView *)mapView
{
    if (!_mapView) {
        _mapView = [[MAMapView alloc] initWithFrame:CGRectMake(0, 0, Screen_Width, kTableViewMaxY())];
        _mapView.showsUserLocation = YES;
        _mapView.userTrackingMode = MAUserTrackingModeFollow;
//        _mapView.language = MAMapLanguageEn;
        _mapView.showsCompass = NO;
        _mapView.showsScale = NO;
        _mapView.zoomLevel = 17;
        _mapView.delegate = self;
    }
    return _mapView;
}

- (AMapSearchAPI *)searchAPI
{
    if (!_searchAPI) {
        _searchAPI = [[AMapSearchAPI alloc]init];
        _searchAPI.delegate = self;
    }
    return _searchAPI;
}

- (UIButton *)myLocationBtn
{
    if (!_myLocationBtn) {
        _myLocationBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [_myLocationBtn setImage:[UIImage imageNamed:TUIKitResource(@"location_my")] forState:UIControlStateNormal];
        [_myLocationBtn addTarget:self action:@selector(myLocationBtnClick) forControlEvents:UIControlEventTouchUpInside];
        _myLocationBtn.frame = CGRectMake(Screen_Width - 60 - 15, kTableViewMaxY() - 60 - 15, 60, 60);
    }
    return _myLocationBtn;
}

- (NSMutableArray<MSLocationInfo *> *)listArr
{
    if (!_listArr) {
        _listArr = [NSMutableArray array];
    }
    return _listArr;
}

- (void)setUserLocation:(CLLocation *)userLocation
{
    _userLocation = userLocation;
    [self addAnnotation:userLocation.coordinate.latitude longitude:userLocation.coordinate.longitude];
}

- (AMapPOIAroundSearchRequest *)poiRequest
{
    if (!_poiRequest) {
        _poiRequest = [[AMapPOIAroundSearchRequest alloc]init];
        _poiRequest.sortrule = 0;
        _poiRequest.requireExtension = YES;
        _poiRequest.types = @"汽车销售|餐饮服务|购物服务|生活服务|医疗保健服务|住宿服务|风景名胜|商务住宅|政府机构及社会团体|科教文化服务|交通设施服务|金融保险服务|地名地址信息|公共设施";
    }
    return _poiRequest;
}

- (void)setupNavView
{
    self.navView = [[UIView alloc]initWithFrame:CGRectMake(0, 0,Screen_Width , StatusBar_Height + NavBar_Height)];
    [self.view addSubview:self.navView];
    [self.navView setGradientLayer:[[UIColor blackColor]colorWithAlphaComponent:0.5] endColor:[[UIColor blackColor]colorWithAlphaComponent:0.0]];
    
    UIButton *cancelBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [cancelBtn setTitle:TUILocalizableString(Cancel) forState:UIControlStateNormal];
    [cancelBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    cancelBtn.titleLabel.font = [UIFont systemFontOfSize:16];
    [cancelBtn addTarget:self action:@selector(cancelBtnClick) forControlEvents:UIControlEventTouchUpInside];
    cancelBtn.frame = CGRectMake(6, StatusBar_Height, 60, NavBar_Height);
    [self.navView addSubview:cancelBtn];
    
    UIButton *sendBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [sendBtn setTitle:TUILocalizableString(Send) forState:UIControlStateNormal];
    [sendBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    sendBtn.titleLabel.font = [UIFont systemFontOfSize:16];
    sendBtn.backgroundColor = [UIColor d_systemBlueColor];
    sendBtn.layer.cornerRadius = 4;
    sendBtn.layer.masksToBounds = YES;
    [sendBtn addTarget:self action:@selector(sendBtnClick) forControlEvents:UIControlEventTouchUpInside];
    sendBtn.frame = CGRectMake(Screen_Width - 10 - 60, StatusBar_Height + 7, 60, 30);
    [self.navView addSubview:sendBtn];
}

- (void)viewDidLayoutSubviews
{
    [super viewDidLayoutSubviews];
    [self.view bringSubviewToFront:self.navView];
}

#pragma mark - button event

- (void)cancelBtnClick
{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)sendBtnClick
{
    if (self.selectInfo == nil) {
        return;
    }
    if (self.selectLocation) {
        self.selectInfo.zoom = self.mapView.zoomLevel;
        self.selectLocation(self.selectInfo);
    }
    [self cancelBtnClick];
}

- (void)myLocationBtnClick
{
    self.userLocation = self.mapView.userLocation.location;
    [self.mapView setCenterCoordinate:self.mapView.userLocation.coordinate animated:YES];
    self.page = 1;
    [self.listArr removeAllObjects];
    [self.myTableView reloadData];
    [self searchPOI];
}

- (void)loadMorePOI
{
    self.page++;
    [self searchPOI];
}

- (void)searchPOI
{
    self.poiRequest.page = self.page;
    self.poiRequest.location = [AMapGeoPoint locationWithLatitude:self.userLocation.coordinate.latitude longitude:self.userLocation.coordinate.longitude];
    [self.searchAPI AMapPOIAroundSearch:self.poiRequest];
}

- (void)addAnnotation:(double)latitude longitude:(double)longitude
{
    [self.mapView removeAnnotation:self.annotation];
    self.annotation = [[MAPointAnnotation alloc] init];
    self.annotation.title = @"pointReuseIndentifier";
    self.annotation.coordinate = CLLocationCoordinate2DMake(latitude, longitude);
    [self.mapView addAnnotation:self.annotation];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.listArr.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    MSLocationListCell *cell = [tableView dequeueReusableCellWithIdentifier:@"locationCell" forIndexPath:indexPath];
    [cell configCell:self.listArr[indexPath.row]];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    self.selectInfo.isSelect = NO;
    MSLocationInfo *info = self.listArr[indexPath.row];
    info.isSelect = YES;
    self.selectInfo = info;
    [self.myTableView reloadData];
    [self.mapView setCenterCoordinate:CLLocationCoordinate2DMake(info.latitude, info.longitude) animated:YES];
    [self addAnnotation:info.latitude longitude:info.longitude];
}

- (void)scrollViewWillBeginDragging:(UIScrollView *)scrollView
{
    CGPoint point = [scrollView.panGestureRecognizer velocityInView:scrollView];
    if (point.y > 0) {//下滑
        if ((NSInteger)self.myTableView.y == (NSInteger)kTableViewMinY() && self.myTableView.contentOffset.y <= 0.f) {
            [UIView animateWithDuration:0.4 animations:^{
                self.myTableView.y = kTableViewMaxY();
                self.myLocationBtn.maxY = kTableViewMaxY() - 15;
                self.mapView.y = 0;
            } completion:^(BOOL finished) {
                self.myTableView.height = Screen_Height - kTableViewMaxY();
            }];
            self.myTableView.scrollEnabled = NO;
        }
    }else {//上滑
        if ((NSInteger)self.myTableView.y == (NSInteger)kTableViewMaxY() && self.myTableView.contentOffset.y <= 0.f) {
            self.myTableView.height = Screen_Height - kTableViewMinY();
            [UIView animateWithDuration:0.4 animations:^{
                self.myTableView.y = kTableViewMinY();
                self.mapView.y = -(kTableViewMaxY() - kTableViewMinY()) * 0.5;
                self.myLocationBtn.maxY = kTableViewMinY() - 15;
            } completion:^(BOOL finished) {
               
            }];
        }
    }
}

- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate
{
    self.myTableView.scrollEnabled = YES;
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView
{
    if (scrollView.contentOffset.y < 0) {
        scrollView.contentOffset = CGPointZero;
    }
}

#pragma mark - MAMapViewDelegate

- (void)mapView:(MAMapView *)mapView didUpdateUserLocation:(MAUserLocation *)userLocation updatingLocation:(BOOL)updatingLocation
{
    if (self.userLocation == nil) {
        self.userLocation = userLocation.location;
        self.page = 1;
        [self.listArr removeAllObjects];
        [self.myTableView reloadData];
        [self searchPOI];
    }
}

- (void)mapView:(MAMapView *)mapView mapDidMoveByUser:(BOOL)wasUserAction
{
    if (wasUserAction) {
        self.page = 1;
        self.userLocation = [[CLLocation alloc]initWithLatitude:mapView.centerCoordinate.latitude longitude:mapView.centerCoordinate.longitude];
        [self.listArr removeAllObjects];
        [self.myTableView reloadData];
        [self searchPOI];
    }
}

- (MAAnnotationView *)mapView:(MAMapView *)mapView viewForAnnotation:(id <MAAnnotation>)annotation
{
    if ([annotation.title isEqualToString:@"pointReuseIndentifier"]) {
        static NSString *pointReuseIndentifier = @"pointReuseIndentifier";
        MAPinAnnotationView *annotationView = (MAPinAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:pointReuseIndentifier];
        if (annotationView == nil) {
            annotationView = [[MAPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:pointReuseIndentifier];
        }
        annotationView.pinColor = MAPinAnnotationColorRed;
        return annotationView;
    }
    return nil;
}

#pragma mark - AMapSearchDelegate

/* POI 搜索回调. */
- (void)onPOISearchDone:(AMapPOISearchBaseRequest *)request response:(AMapPOISearchResponse *)response
{
    [self.myTableView.mj_footer endRefreshing];
    if (response.pois.count == 0) {
        [self.myTableView.mj_footer endRefreshingWithNoMoreData];
    }
    for (AMapPOI *poi in response.pois) {
        MSLocationInfo *info = [[MSLocationInfo alloc]init];
        info.name = poi.name;
        info.address = poi.address;
        info.detail = [NSString stringWithFormat:@"%@%@%@%@",XMNoNilString(poi.province),XMNoNilString(poi.city),XMNoNilString(poi.district),XMNoNilString(poi.address)];
        info.distance = poi.distance;
        info.city = poi.city;
        info.province = poi.province;
        info.district = poi.district;
        info.latitude = poi.location.latitude;
        info.longitude = poi.location.longitude;
        [self.listArr addObject:info];
    }
    if (self.page == 1) {
        self.selectInfo = self.listArr.firstObject;
        self.selectInfo.isSelect = YES;
    }
    [self.myTableView reloadData];
}

- (void)AMapSearchRequest:(id)request didFailWithError:(NSError *)error
{
    MSLog(@"AMapSearchRequest failed: %@",error);
    [self.myTableView.mj_footer endRefreshing];
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/Location/MSLocationController.swift
//
//  MSLocationController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/12/9.
//

import UIKit
import MJRefresh
import MAMapKit
import AMapFoundationKit
import AMapSearchKit
import MSIMSDK


class MSLocationController: UIViewController {

    /// 点击发送位置回调
    var didSendLocation: ((_ info: MSLocationInfo) -> Void)?
    
    private var listArr: [MSLocationInfo] = []
    
    private var userLocation: CLLocation? {
        didSet {
            if userLocation != nil {
                self.addAnnotation(latitude: userLocation!.coordinate.latitude, longitude: userLocation!.coordinate.longitude)
            }
        }
    }
    
    private var selectInfo: MSLocationInfo?
    
    private var page: Int = 1
    
    private var annotation: MAPointAnnotation?
    
    lazy private var myTableView: UITableView = {
        let myTableView = UITableView(frame: CGRect(x: 0, y: kTableViewMaxY, width: UIScreen.width, height: UIScreen.height - kTableViewMaxY), style: .plain)
        myTableView.delegate = self
        myTableView.dataSource = self
        myTableView.tableFooterView = UIView()
        myTableView.backgroundColor = self.view.backgroundColor
        myTableView.register(MSLocationListCell.self, forCellReuseIdentifier: "locationCell")
        myTableView.rowHeight = 70
        myTableView.separatorColor = UIColor.d_color(light: MSMcros.TCell_separatorColor, dark: MSMcros.TCell_separatorColor_Dark)
        myTableView.layer.cornerRadius = 16
        myTableView.clipsToBounds = true
        myTableView.mj_footer = MJRefreshAutoNormalFooter(refreshingTarget: self, refreshingAction: #selector(loadMorePOI))
        return myTableView
    }()

    private let kTableViewMaxY = UIScreen.height * 0.6
    private let kTableViewMinY = UIScreen.height * 0.4

    lazy private var mapVeiw: MAMapView = {
        let mapView = MAMapView(frame: CGRect(x: 0, y: 0, width: UIScreen.width, height: kTableViewMaxY))
        mapView.showsUserLocation = true
        mapView.userTrackingMode = .follow
        mapView.showsCompass = false
        mapView.showsScale = false
        mapView.zoomLevel = 17
        mapView.delegate = self
        return mapView
    }()
    
    lazy private var searchAPI: AMapSearchAPI = {
        let api = AMapSearchAPI()!
        api.delegate = self
        return api
    }()
    
    lazy private var myLocationBtn: UIButton = {
        let btn = UIButton(type: .custom)
        btn.setImage(UIImage.bf_imageNamed(name: "location_my"), for: .normal)
        btn.addTarget(self, action: #selector(myLocationBtnClick), for: .touchUpInside)
        btn.frame = CGRect(x: UIScreen.width - 60 - 15, y: kTableViewMaxY - 60 - 15, width: 60, height: 60)
        return btn
    }()
    
    lazy private var poiRequest: AMapPOIAroundSearchRequest = {
        let request = AMapPOIAroundSearchRequest()
        request.sortrule = 0
        request.requireExtension = true
        request.types = "汽车销售|餐饮服务|购物服务|生活服务|医疗保健服务|住宿服务|风景名胜|商务住宅|政府机构及社会团体|科教文化服务|交通设施服务|金融保险服务|地名地址信息|公共设施"
        return request
    }()
    
    lazy private var navView: UIView = {
        let navView = UIView(frame: CGRect(x: 0, y: 0, width: UIScreen.width, height:UIScreen.status_navi_height))
        return navView
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor.d_color(light: MSMcros.TController_Background_Color, dark: MSMcros.TController_Background_Color_Dark)
        
        setupNavView()
        AMapServices.shared().apiKey = MSMcros.GaodeAPIKey
        AMapSearchAPI.updatePrivacyShow(.didShow, privacyInfo: .didContain)
        AMapSearchAPI.updatePrivacyAgree(.didAgree)
        AMapServices.shared().enableHTTPS = true
        
        view.addSubview(mapVeiw)
        view.addSubview(myTableView)
        view.addSubview(myLocationBtn)
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        view.bringSubviewToFront(self.navView)
    }
    
    deinit {
        print("%@ dealloc",self)
    }
}

extension MSLocationController: UITableViewDataSource,UITableViewDelegate {
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return listArr.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "locationCell", for: indexPath) as! MSLocationListCell
        cell.locationInfo = listArr[indexPath.row]
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        self.selectInfo?.isSelect = false
        let info = self.listArr[indexPath.row]
        info.isSelect = true
        self.selectInfo = info
        self.myTableView.reloadData()
        self.mapVeiw.setCenter(CLLocationCoordinate2DMake(info.latitude, info.longitude), animated: true)
        self.addAnnotation(latitude: info.latitude, longitude: info.longitude)
    }
}

extension MSLocationController {
    
    // MARK: --- private event
    
    private func setupNavView() {
        view.addSubview(navView)
        navView.addGradientLayer(startColor: UIColor.black.withAlphaComponent(0.5), endColor: UIColor.black.withAlphaComponent(0.0))
        
        let cancelBtn = UIButton(type: .custom)
        cancelBtn.setTitle(Bundle.bf_localizedString(key: "Cancel"), for: .normal)
        cancelBtn.setTitleColor(.white, for: .normal)
        cancelBtn.titleLabel?.font = .systemFont(ofSize: 16)
        cancelBtn.addTarget(self, action: #selector(cancelBtnClick), for: .touchUpInside)
        cancelBtn.frame = CGRect(x: 6, y: UIScreen.statusbarHeight, width: 60, height: UIScreen.navBarHeight)
        navView.addSubview(cancelBtn)
        
        let sendBtn = UIButton(type: .custom)
        sendBtn.setTitle(Bundle.bf_localizedString(key: "Send"), for: .normal)
        sendBtn.setTitleColor(.white, for: .normal)
        sendBtn.titleLabel?.font = .systemFont(ofSize: 16)
        sendBtn.backgroundColor = .systemBlue
        sendBtn.layer.cornerRadius = 4
        sendBtn.layer.masksToBounds = true
        sendBtn.addTarget(self, action: #selector(SendBtnClick), for: .touchUpInside)
        sendBtn.frame = CGRect(x: UIScreen.width - 10 - 60, y: UIScreen.statusbarHeight + 7, width: 60, height: 30)
        navView.addSubview(sendBtn)
    }
    
    @objc func loadMorePOI() {
        self.page += 1
        
    }
    
    @objc func myLocationBtnClick() {
        self.userLocation = self.mapVeiw.userLocation.location
        self.mapVeiw.setCenter(self.mapVeiw.userLocation.coordinate, animated: true)
        self.page = 1
        self.listArr.removeAll()
        self.myTableView.reloadData()
        self.searchPOI()
    }
    
    @objc func cancelBtnClick() {
        self.dismiss(animated: true, completion: nil)
    }
    
    @objc func SendBtnClick() {
        if self.selectInfo == nil {return}
        if self.didSendLocation != nil {
            self.selectInfo?.zoom = Int(self.mapVeiw.zoomLevel)
            self.didSendLocation?(self.selectInfo!)
        }
        self.cancelBtnClick()
    }
    
    private func searchPOI() {
        self.poiRequest.page = self.page
        if self.userLocation != nil {
            self.poiRequest.location = AMapGeoPoint.location(withLatitude: self.userLocation!.coordinate.latitude, longitude: self.userLocation!.coordinate.longitude)
            self.searchAPI.aMapPOIAroundSearch(self.poiRequest)
        }
    }
    
    private func addAnnotation(latitude: Double,longitude: Double) {
        if self.annotation != nil {
            self.mapVeiw.removeAnnotation(self.annotation!)
        }
        self.annotation = MAPointAnnotation()
        self.annotation?.title = "pointReuseIndentifier"
        self.annotation?.coordinate = CLLocationCoordinate2DMake(latitude, longitude)
        self.mapVeiw.addAnnotation(self.annotation!)
    }
}

// MARK: MAMapViewDelegate
extension MSLocationController: MAMapViewDelegate {
    
    func mapView(_ mapView: MAMapView!, didUpdate userLocation: MAUserLocation!, updatingLocation: Bool) {
        if self.userLocation == nil {
            self.userLocation = userLocation.location
            self.page = 1
            self.listArr.removeAll()
            self.myTableView.reloadData()
            self.searchPOI()
        }
    }
    
    func mapView(_ mapView: MAMapView!, mapDidMoveByUser wasUserAction: Bool) {
        if wasUserAction {
            self.page = 1
            self.userLocation = CLLocation(latitude: mapView.centerCoordinate.latitude, longitude: mapView.centerCoordinate.longitude)
            self.listArr.removeAll()
            self.myTableView.reloadData()
            self.searchPOI()
        }
    }
    
    func mapView(_ mapView: MAMapView!, viewFor annotation: MAAnnotation!) -> MAAnnotationView? {
        if annotation.title == "pointReuseIndentifier" {
            let pointReuseIndentifier = "pointReuseIndentifier"
            if let annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: pointReuseIndentifier) as? MAPinAnnotationView {
                annotationView.pinColor = .red
                return annotationView
            }
            let annotationView = MAPinAnnotationView(annotation: annotation, reuseIdentifier: pointReuseIndentifier)!
            annotationView.pinColor = .red
            return annotationView
        }
        return nil
    }
}

// MARK: AMapSearchDelegate
extension MSLocationController: AMapSearchDelegate {
    
    func onPOISearchDone(_ request: AMapPOISearchBaseRequest!, response: AMapPOISearchResponse!) {
        
        self.myTableView.mj_footer?.endRefreshing()
        if response.pois.count == 0 {
            self.myTableView.mj_footer?.endRefreshingWithNoMoreData()
        }
        for poi in response.pois {
            
            let info = MSLocationInfo(poi: poi)
            self.listArr.append(info)
        }
        if self.page == 1 {
            self.selectInfo = self.listArr.first
            self.selectInfo?.isSelect = true
        }
        self.myTableView.reloadData()
    }
       
    func aMapSearchRequest(_ request: Any!, didFailWithError error: Error!) {
        print("AMapSearchRequest failed: \(String(describing: error))")
        self.myTableView.mj_footer?.endRefreshing()
    }
}

extension MSLocationController {
    
    func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
        let point = scrollView.panGestureRecognizer.velocity(in: scrollView)
        if point.y > 0 {//下滑
            
            if self.myTableView.top == kTableViewMinY && self.myTableView.contentOffset.y <= 0 {
                UIView.animate(withDuration: 0.4) {
                    self.myTableView.top = self.kTableViewMaxY
                    self.myLocationBtn.bottom = self.kTableViewMaxY - 15
                    self.mapVeiw.top = 0
                } completion: { _ in
                    self.myTableView.height = UIScreen.height - self.kTableViewMaxY
                }
                self.myTableView.isScrollEnabled = false
            }
        }else {//上滑
            if self.myTableView.top == kTableViewMaxY && self.myTableView.contentOffset.y <= 0 {
                self.myTableView.height = UIScreen.height - kTableViewMinY
                UIView.animate(withDuration: 0.4) {
                    self.myTableView.top = self.kTableViewMinY
                    self.mapVeiw.top = -(self.kTableViewMaxY - self.kTableViewMinY) * 0.5
                    self.myLocationBtn.bottom = self.kTableViewMinY - 15
                } completion: { _ in
                    
                }

            }
        }
    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        self.myTableView.isScrollEnabled = true
    }
    
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        if scrollView.contentOffset.y < 0 {
            scrollView.contentOffset = .zero
        }
    }
}
